enclave-runner/src/usercalls/mod.rs

```diff
@@ -346,37 +385,228 @@ impl EnclaveState {
+                                #[cfg(unix)]
+                                trap_attached_debugger(usercall.tcs.address().0 as _);
+                                my_result = Err(EnclaveAbort::Exit {
+                                    panic: EnclavePanic::from(buf.into_inner()),
+                                });
+                            },
+                            Err(EnclaveAbort::Exit { panic: false }) => {
+                                if state.mode != EnclaveEntry::ExecutableMain {
+                                    erased_tcs = Some(usercall.tcs);
+                                }
+                                my_result = Ok((0,0));
+                            },
+                            Err(EnclaveAbort::IndefiniteWait) => my_result = Err(EnclaveAbort::IndefiniteWait),
+                            Err(EnclaveAbort::InvalidUsercall(n)) => my_result = Err(EnclaveAbort::InvalidUsercall(n)),
+                            Err(EnclaveAbort::MainReturned) => my_result = Err(EnclaveAbort::MainReturned),
+                            Err(EnclaveAbort::Secondary) => my_result = Err(EnclaveAbort::Secondary),
+                        }
+                    },
+                    CoResult::Return((tcs,v1, v2)) => {
+                        match state.mode {
+                            EnclaveEntry::Library => {
+                                erased_tcs = Some(tcs);
+                                my_result = Ok((v1, v2));
+                            }
+                            EnclaveEntry::ExecutableMain => {
+                                main_result = Err(EnclaveAbort::MainReturned);
+                                break 'outer;
+                            }
+                            EnclaveEntry::ExecutableNonMain => {
+                                assert_eq!(
+                                    (v1, v2),
+                                    (0, 0),
+                                    "Expected enclave thread entrypoint to return zero"
+                                );
+                                erased_tcs = Some(tcs);
+                                my_result = Ok((0,0));
+                            }
+                        }
+                    }
+                };
+                if state.mode != EnclaveEntry::ExecutableNonMain && my_result.is_err() {
+                    main_result = my_result;
+                    break 'outer;
+                }
+                match my_result {
+                    Ok(ret) => {
+                        if state.mode == EnclaveEntry::ExecutableNonMain {
+                            let cmd = enclave.kind.as_command().unwrap();
+                            let mut cmddata = cmd.data.lock().unwrap();
+                            cmddata.running_secondary_threads -= 1;
+                            if cmddata.running_secondary_threads == 0 {
+                                cmd.wait_secondary_threads.notify_all();
```

Comment:
The only thing that could be waiting is the main thread, but this is the main thread and it's executing this. As I mentioned in my comments earlier this week, I'm pretty sure this notification system isn't going to work anymore.

The point of the notification system was to make sure that `Command::run` didn't return as long as any enclave thread was still executing. Now that we have the worker thread system, you can do this differently. I think you can just join all the worker threads before returning from main_entry. I believe worker threads should be exiting automatically after the event loop is closed (because all the work sender channels are closed).